### PR TITLE
chore: Upgrade to `pnpm@12` (and update Turborepo)

### PR DESCRIPTION
**Pull request by kitten**
# Why

This is to upgrade us to `pnpm@12` (which is basically a no-change major update compared to v11, but with a native resolution module, as far as I'm aware)

There's a few changes in here, some are beneficial for us (catching a few common mistakes), some are adverse to us (moving a lot of configs to `pnpm-workspace.yaml` instead of allowing some configs in `package.json`)

This is a prerequisite to automating releases, and the upgrade is opportune, since it roots out a few inconsistencies. However, it does have a few drawbacks.

Prior PRs that were prerequisites for this one are:
- #49646
- #49628
- #49615

# How

The main update to Turborepo is to see if it resolves the sporadic Windows failures. It might not, but a few other patches here are interesting for us. However, it does regress on workspace detection for `pnpm@10`, which causes two errors. But those are resolved by adding `package.json:devEngines.packageManager`

The update to pnpm causes a few other changes:
- Many `package.json` options like `resolutions` are ignored
- `npm_config_*` overrides are ignored
- Patches may be rejected if they result in no change
- `CI=1` consistently enforces frozen lockfile unless overridden
- Peer resolution is now slightly different

The last point led to an issue with `debug` which has an optional peer on `supports-color`. However, due to how `supports-color` is pulled in in our dependencies, and `debug` being used in many places, this split our dependency graph in half due to peer changes in `debug` cascading upwards. This is an edge case and handled using a deps override.

There's unfortunately a lot of churn in various E2E tests, since the remaining ones that don't have a stable workspace create adhoc projects or adhoc modify the monorepo, causing various failures due to assumptions on package manager semantics and behaviour. This has been a bit hard to narrow down, but the priority is to get this merged first and to address these E2E tests separately.

> [!NOTE]
> As each commit addresses a separate problem past the first one, it's better to step through those to see the full list of changes. Some of these are workarounds, since these downstream problems otherwise become a blocker here

### CODE REVIEW

File: tools/src/commands/BumpReactNativeVersion.ts [hunk 1]
```diff
@@ -145,36 +144,38 @@ async function updatePackageJson(packageJsonPath: string, newVersion: string): P
...
   }
 
-  if (modified) {
-    await JsonFile.writeAsync(ROOT_PACKAGE_JSON_PATH, json);
-    logger.log(`  Updated ${chalk.cyan('package.json')} (resolutions)`);
-  }
+  if (!modified) return false;
 
-  return modified;
+  await spawnAsync(
```
Comment (kitten) [inline tools/src/commands/BumpReactNativeVersion.ts:172]: note: I'd treat this command/script as semi-temporary, since pnpm catalogs should/can take its place more reliably, presumably. There's maybe still a flow we can capture, but I suspect it'll change this script quite a bit. Out of scope for this PR though

File: bin/expo-turbo [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Comment (kitten) [inline bin/expo-turbo:1]: This is a workaround for a Turborepo regression. In some cases, it fails to treat the root `package.json` as such, and assumes it's a regular workspace. This means that it doesn't allow `turbo` commands to be added as scripts there, like for workspace `package.json`s.

To work around this, while we're preserving the `turbo build` invocation in `prepare`, and while we're not fully migrated, this wrapper script works around this error

**Review by hassankhan — APPROVED** [review]
